File: rlm/cli/wizard/channels.py
"""Especificações e testes de canais de comunicação (Telegram, Discord, etc.)."""

from __future__ import annotations

from typing import Any, Callable, TypedDict

ChannelVarSpec = tuple[str, str, bool]
ChannelTestFn = Callable[[str], tuple[bool, str]]
_TEST_REQUEST_TIMEOUT_S = 10


class ChannelSpec(TypedDict):
    name: str
    id: str
    vars: list[ChannelVarSpec]
    hint: str
    test_fn: ChannelTestFn | None
    test_env_var: str | None


def _redact_secret(text: str, secret: str) -> str:
    if not secret:
        return text
    return text.replace(secret, "***")


def _http_json_get(
    url: str,
    *,
    headers: dict[str, str] | None = None,
    redact_secret: str = "",
) -> tuple[dict[str, Any] | None, str | None]:
    import json as _json
    from urllib import error as _urlerr
    from urllib import request as _urlreq

    req = _urlreq.Request(url, headers=headers or {}, method="GET")
    try:
        with _urlreq.urlopen(req, timeout=_TEST_REQUEST_TIMEOUT_S) as resp:
            payload = _json.loads(resp.read().decode("utf-8"))
        return payload, None
    except _urlerr.HTTPError as exc:
        return None, f"HTTP {exc.code}"
    except Exception as exc:
        return None, _redact_secret(str(exc), redact_secret)


def is_sensitive_channel_var(var_name: str) -> bool:
    upper_name = var_name.upper()
    return any(token in upper_name for token in ("TOKEN", "SECRET", "PASSWORD", "API_KEY"))


def _test_telegram_token(token: str) -> tuple[bool, str]:
    """Testa token do Telegram via /getMe. Retorna (ok, mensagem)."""
    url = f"https://api.telegram.org/bot{token}/getMe"
    data, error = _http_json_get(url, redact_secret=token)
    if error is not None:
        return False, error
    assert data is not None
    if data.get("ok"):
        bot = data.get("result", {})
        name = bot.get("first_name", "?")
        uname = bot.get("username", "?")
        return True, f"@{uname} ({name})"
    return False, "API retornou ok=false"


def _test_discord_token(token: str) -> tuple[bool, str]:
    """Testa token do Discord via /users/@me. Retorna (ok, mensagem)."""
    url = "https://discord.com/api/v10/users/@me"
    data, error = _http_json_get(
        url,
        headers={"Authorization": f"Bot {token}"},
        redact_secret=token,
    )
    if error is not None:
        return False, error
    assert data is not None
    uname = data.get("username", "?")
    return True, f"@{uname}"


_CHANNEL_SPECS: list[ChannelSpec] = [
    {
        "name": "Telegram",
        "id": "telegram",
        "vars": [
            ("TELEGRAM_BOT_TOKEN", "Bot Token (do @BotFather)", True),
            ("TELEGRAM_OWNER_CHAT_ID", "Chat ID do dono (para notificações)", False),
        ],
        "hint": "Converse com @BotFather -> /newbot -> copie o token",
        "test_fn": _test_telegram_token,
        "test_env_var": "TELEGRAM_BOT_TOKEN",
    },
    {
        "name": "Discord",
        "id": "discord",
        "vars": [
            ("DISCORD_BOT_TOKEN", "Bot Token", False),
            ("DISCORD_APP_PUBLIC_KEY", "Public Key (Ed25519)", True),
            ("DISCORD_APP_ID", "Application ID", True),
        ],
        "hint": "Discord Developer Portal -> Applications -> Bot -> Token",
        "test_fn": _test_discord_token,
        "test_env_var": "DISCORD_BOT_TOKEN",
    },
    {
        "name": "WhatsApp",
        "id": "whatsapp",
        "vars": [
            ("WHATSAPP_TOKEN", "Access Token (Meta Cloud API)", True),
            ("WHATSAPP_PHONE_ID", "Phone Number ID", True),
            ("WHATSAPP_VERIFY_TOKEN", "Webhook Verify Token (defina você)", True),
        ],
        "hint": "Meta for Developers -> Your App -> WhatsApp -> Configuration",
        "test_fn": None,
        "test_env_var": None,
    },
    {
        "name": "Slack",
        "id": "slack",
        "vars": [
            ("SLACK_BOT_TOKEN", "Bot User OAuth Token (xoxb-...)", True),
            ("SLACK_SIGNING_SECRET", "Signing Secret", True),
        ],
        "hint": "Slack API -> Your App -> OAuth & Permissions + Basic Information",
        "test_fn": None,
        "test_env_var": None,
    },
]

CHANNEL_SPECS = _CHANNEL_SPECS

File: packages/cli/src/wizard/channels.ts
/**
 * Especificações e testes de canais de comunicação (Telegram, Discord, etc.).
 *
 * Migrado de rlm/cli/wizard/channels.py
 */

// ─────────────────────────────────────────────── Tipos

export interface ChannelVarSpec {
  /** Nome da variável de ambiente */
  key: string;
  /** Rótulo para exibição no wizard */
  label: string;
  /** Se verdadeiro, o wizard alerta sobre ausência */
  required: boolean;
}

export interface ChannelSpec {
  name: string;
  id: string;
  vars: ChannelVarSpec[];
  hint: string;
  /** Nome interno da função de teste, se houver */
  testFn?: string;
}

// ─────────────────────────────────────────────── Especificações dos canais

export const CHANNEL_SPECS: ChannelSpec[] = [
  {
    name: "Telegram",
    id: "telegram",
    vars: [
      { key: "TELEGRAM_BOT_TOKEN", label: "Bot Token (do @BotFather)", required: true },
      { key: "TELEGRAM_OWNER_CHAT_ID", label: "Chat ID do dono (para notificações)", required: false },
    ],
    hint: "Converse com @BotFather → /newbot → copie o token",
    testFn: "testTelegramToken",
  },
  {
    name: "Discord",
    id: "discord",
    vars: [
      { key: "DISCORD_BOT_TOKEN", label: "Bot Token", required: false },
      { key: "DISCORD_APP_PUBLIC_KEY", label: "Public Key (Ed25519)", required: true },
      { key: "DISCORD_APP_ID", label: "Application ID", required: true },
    ],
    hint: "Discord Developer Portal → Applications → Bot → Token",
    testFn: "testDiscordToken",
  },
  {
    name: "WhatsApp",
    id: "whatsapp",
    vars: [
      { key: "WHATSAPP_TOKEN", label: "Access Token (Meta Cloud API)", required: true },
      { key: "WHATSAPP_PHONE_ID", label: "Phone Number ID", required: true },
      { key: "WHATSAPP_VERIFY_TOKEN", label: "Webhook Verify Token (defina você)", required: true },
    ],
    hint: "Meta for Developers → Your App → WhatsApp → Configuration",
  },
  {
    name: "Slack",
    id: "slack",
    vars: [
      { key: "SLACK_BOT_TOKEN", label: "Bot User OAuth Token (xoxb-…)", required: true },
      { key: "SLACK_SIGNING_SECRET", label: "Signing Secret", required: true },
    ],
    hint: "Slack API → Your App → OAuth & Permissions + Basic Information",
  },
];

// ─────────────────────────────────────────────── Funções de teste

/**
 * Testa token do Telegram via /getMe.
 * Retorna [ok, mensagem descritiva].
 */
export async function testTelegramToken(token: string): Promise<[boolean, string]> {
  const url = `https://api.telegram.org/bot${encodeURIComponent(token)}/getMe`;
  try {
    const resp = await fetch(url, {
      method: "GET",
      signal: AbortSignal.timeout(10_000),
    });
    if (!resp.ok) {
      return [false, `HTTP ${resp.status}`];
    }
    const data = (await resp.json()) as Record<string, unknown>;
    if (data["ok"]) {
      const result = (data["result"] ?? {}) as Record<string, unknown>;
      const name = String(result["first_name"] ?? "?");
      const uname = String(result["username"] ?? "?");
      return [true, `@${uname} (${name})`];
    }
    return [false, "API retornou ok=false"];
  } catch (err) {
    return [false, String(err)];
  }
}

/**
 * Testa token do Discord via /users/@me.
 * Retorna [ok, mensagem descritiva].
 */
export async function testDiscordToken(token: string): Promise<[boolean, string]> {
  const url = "https://discord.com/api/v10/users/@me";
  try {
    const resp = await fetch(url, {
      method: "GET",
      headers: { Authorization: `Bot ${token}` },
      signal: AbortSignal.timeout(10_000),
    });
    if (!resp.ok) {
      return [false, `HTTP ${resp.status}`];
    }
    const data = (await resp.json()) as Record<string, unknown>;
    const uname = String(data["username"] ?? "?");
    return [true, `@${uname}`];
  } catch (err) {
    return [false, String(err)];
  }
}

/** Mapa de funções de teste por nome interno. */
export const CHANNEL_TEST_FNS: Record<
  string,
  (token: string) => Promise<[boolean, string]>
> = {
  testTelegramToken,
  testDiscordToken,
};
